spec/index.bs

```diff
@@ -1191,131 +1212,189 @@ access to it before garbage collection by calling {{GPUBuffer/destroy()}}.
-        1. Return [=a promise rejected with=] an {{AbortError}}.
+        1. Set |this|.{{[[mapping]]}} to |m|.
+        1. Let |m| be a new {{ArrayBuffer}} of size |size|.
+        1. Set the content of |m| to the content of |this|'s allocation starting at offset |offset| and for |size| bytes.
+        1. Set |this|.{{[[state]]}} to [=buffer state/mapped=].
+        1. Set |this|.{{[[mapping_range]]}} to `[start, offset]`.
+        1. Set |this|.{{[[mapped_ranges]]}} to `[]`.
+        1. Resolve |p|.
 
-        Issue: Specify that the rejection happens on the device timeline.
+  1. Return |p|.
 
-    1. Let |p| be a new {{Promise}}.
-    1. Set the {{[[mapping]]}} slot of |this| to |p|.
-    1. Set the {{[[state]]}} slot of |this| to [=buffer state/mapping pending for reading=].
-    1. Enqueue an operation on the [=Queue timeline=] that will execute the following:
+  <div algorithm class=validusage>
+    <dfn abstract-op>mapAsync Valid Usage</dfn>
 
-        1. Let |m| be a new {{ArrayBuffer}} of size the {{[[size]]}} of |this|.
-        1. Set the content of |m| to the content of |this|'s allocation.
-        1. Set the {{[[state]]}} slot of |this| to [=buffer state/mapped for reading=].
-        1. If |p| is pending:
+      Given a {{GPUBuffer}} |this|, a {{GPUSize64}} |offset| and a {{GPUSize64}} |size|
+      the following validation rules apply:
 
-            1. Resolve |p| with |m|.
+      1, |this| must be a [=valid=] {{GPUBuffer}}.
+      1. |offset| must be a multiple of 4.
+      1. |size| must be a multiple of 4.
+      1. |offset| + |size| must be less or equal to |this|.{{[[size]]}}
+      1. |this|.{{[[usage]]}} must contain {{GPUBufferUsage/MAP_READ}} or {{GPUBufferUsage/MAP_WRITE}}.
+      1. |this|.{{[[state]]}} must be [=buffer state/unmapped=]
 
-    1. Return |p|.
+    </dfn>
+  </div>
 </div>
 
-### {{GPUBuffer/mapWriteAsync|GPUDevice.mapWriteAsync}} ### {#GPUBuffer-mapWriteAsync}
+### <dfn method for=GPUBuffer>getMappedRange(start, offset)</dfn> ### {#GPUBuffer-getMappedRange}
 
-<div algorithm="GPUBuffer.mapWriteAsync">
+<div algorithm="GPUBuffer.getMappedRange">
 
-    Issue: Handle error buffers once we have a description of the error monad.
+  **Arguments:**
+    - {{GPUSize64}} |offset|
+    - {{GPUSize64}} |size|
+
+  **Returns:** {{ArrayBuffer}}
 
-    1. If the {{[[usage]]}} slot of |this| doesn't contain the {{GPUBufferUsage/MAP_WRITE}} bit or
-        if {{[[state]]}} isn't [=buffer state/unmapped=]:
+  1. If this call doesn't follow the [$getMappedRange Valid Usage$]:
 
-        1. Record a validation error on the current scope.
-        1. Return [=a promise rejected with=] an {{AbortError}}.
+    1. Throw an {{OperationError}}.
 
-        Issue: Specify that the rejection happens on the device timeline.
+  1. Let |m| be a new {{ArrayBuffer}} of size |size| pointing at the content of |this|.{{[[mapping]]}} at offset |offset| - |this|.{{[[mapping_range]]}}[0].
```

Comment:
```suggestion
  1. Let |m| be a new {{ArrayBuffer}} of size |size| pointing at the content of |this|.{{[[mapping]]}} at offset |offset|.
```
Right? If both `|offset|` and `|this|.{{[[mapping_range]]}}[0]` are relative to the start of the `|this|.{{[[mapping]]}}`?

Reply:
This is confusing, but I think the original text is correct because `[[mapping]]` was created in the following line:

> Set the content of `m` to the content of `this`’s allocation starting at offset `offset` and for `size` bytes.
>
> Set `this.[[mapping]]` to `m`.

`[[mapping_range]]` is relative to the whole allocation of `this`, same for `offset` and `size` of `getMappedRange`, but `[[mapping]]` is a subrange of the allocation. This is all very confusing and I'd be happy to reword things if you find clearer names / explanations.